### PR TITLE
refactor: Remove unused memory introspection functions

### PR DESCRIPTION
**Pull request by kattrali**
## Goal

Slim the codebase by removing functions for which the output is unused. Only string contents are later analyzed for usefulness in reports.

## Design

PR https://github.com/bugsnag/bugsnag-cocoa/pull/188 added support reading and appending Swift error messages to reports by reading register contents and using a heuristic to find likely message contents. This change enabled the `introspectMemory` configuration option, which enabled the flow to inspect and read what is currently in memory:

```
                                       ┌───────────────┐                                         
                                       │ Write Report  │                                         
                                       └───────────────┘                                         
                                               │                                                 
                                               ▼                                                 
                                  ┌────────────────────────┐                                     
                                  │  Memory Introspection  │                                     
                                  │        Enabled?        │                                     
                                  └────────────────────────┘                                     
                                               │                                                 
                                               ▼                                                 
                                   ┌───────────────────────┐                                     
                                   │Read register contents │                                     
                                   │                       │                                     
                                   └───────────────────────┘                                     
                                               │                                                 
         ┌──────────────────┬──────────────────┼──────────────────┬───────────────────┐          
         ▼                  ▼                  ▼                  ▼                   ▼          
┌────────────────┐ ┌────────────────┐ ┌────────────────┐ ┌────────────────┐ ┌───────────────────┐
│ Is it a Date?  │ │  Is it a URL?  │ │Is it a Number? │ │Is it a String? │ │Is it a char array?│
└────────────────┘ └────────────────┘ └────────────────┘ └────────────────┘ └───────────────────┘
         │                  │                  │                 │                    │          
         └──────────────────┴──────────────────┼─────────────────┴────────────────────┘          
                                               ▼                                                 
                                   ┌──────────────────────┐                                      
                                   │    Write register    │                                      
                                   │       contents       │                                      
                                   └──────────────────────┘                                      
```

However in all cases the message and error class of the report are being read from static strings ([which are char arrays under the hood](https://github.com/apple/swift/blob/123f4b90939bc62646aeefa0add3cc8a078957a5/stdlib/public/core/StaticString.swift#L22-L26)) recorded from [`_assertionFailure()` calls](https://github.com/apple/swift/blob/123f4b90939bc62646aeefa0add3cc8a078957a5/stdlib/public/core/Assert.swift#L133). A lot of extra work was being done to introspect other types of pointers unnecessarily at crash time.

This change optimizes the flow to check if a pointer is not a likely object and further is a likely char array, and then only in that case records the value.

```
     ┌───────────────┐    
     │ Write Report  │    
     └───────────────┘    
             │            
             ▼            
┌────────────────────────┐
│  Memory Introspection  │
│        Enabled?        │
└────────────────────────┘
             │            
             ▼            
 ┌───────────────────────┐
 │Read register contents │
 └───────────────────────┘
             │            
             ▼            
    ┌────────────────┐    
    │  Is it not an  │    
    │    object?     │    
    └────────────────┘    
             │            
             ▼            
    ┌────────────────┐    
    │  Is it a char  │    
    │     array?     │    
    └────────────────┘    
             │            
             ▼            
 ┌──────────────────────┐ 
 │    Write register    │ 
 │       contents       │ 
 └──────────────────────┘ 
```

Optimizing this flow made a lot of functions become unused, so the iterations in the commits in these changesets are intended to remove them in an iterative way. This was a very easy process with Warnings as Errors enabled, and an unused function issuing a warning. The biggest wins here being a significant reduction in work performed at crash time and an 18% reduction in the size of the codebase (nearly 4kLOC).

## Tests

* I tested this change by manually running a crash suite and verifying the changes on iOS 11 & 12 and macOS 10.14. 
* The increase in the coverage of the test matrix is intended to further increase confidence going forward.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,22 +4,33 @@ cache:
   - cocoapods
 matrix:
     include:
+        - osx_image: xcode11
```
[the remark this answers is not in the document]
Comment (Cawllec) [reply .travis.yml]: This looks OK to me.  If it's just the unit tests then running it as built by xcode11 as well as xcode10.2 is probably a good idea.

File: Source/BugsnagCrashReport.m [hunk 2]
```diff
@@ -705,9 +705,6 @@ + (instancetype)errorDataFromThreads:(NSArray *)threads {
         NSString *reservedWord = nil;
 
         for (NSString *key in notableAddresses) {
-            if ([key hasPrefix:@"stack"]) { // skip stack frames, only use register values
```
[the remark this answers is not in the document]
Comment (kattrali) [reply Source/BugsnagCrashReport.m]: Yes. This check was needed to exclude values which were never used, and instead this change is a companion to a change to [no longer write the stack contents at all](https://github.com/bugsnag/bugsnag-cocoa/pull/387/commits/0920f5069f41afd28ff0f5321bff10f8954f4a19#diff-3191f957f02d7186d9807ad19f1f933eL925).

File: Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashContext.h [hunk 3]
```diff
@@ -48,13 +48,6 @@ typedef struct {
...
      */
     bool enabled;
-
-    /** List of classes that should never be introspected.
-     * Whenever a class in this list is encountered, only the class name will be
-     * recorded.
-     */
-    const char **restrictedClasses;
-    size_t restrictedClassesCount;
 } BSG_KSCrash_IntrospectionRules;
```
[the remark this answers is not in the document]
Comment (kattrali) [reply Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashContext.h]: Correct. The restricted classes were always set to an empty collection, so they could be removed altogether.

**Comment by kattrali** [comment]
Added a bit more explanation in the description.

**Pull request merged**